pelias-geocoder.js

```diff
@@ -8,294 +13,429 @@ function PeliasGeocoder(opts) {
   this.opts.useFocusPoint = opts.useFocusPoint;
   this.opts.removeDuplicates = opts.removeDuplicates === undefined ? true : opts.removeDuplicates;
   this.opts.onSubmitOnly = opts.onSubmitOnly;
+
   if (opts.marker) {
     this.opts.marker = {};
     this.opts.marker.icon = opts.marker.icon || 'marker-15';
     this.opts.marker.anchor = opts.marker.anchor || 'bottom';
+    this.opts.marker.multiple = opts.marker.multiple !== undefined ? opts.marker.multiple : true;
+    this._customHtmlMarkers = [];
   }
+
   if (opts.wof) {
-    this.opts.wof = {}
+    this.opts.wof = {};
     this.opts.wof.url = opts.wof.url || 'https://raw.githubusercontent.com/whosonfirst-data/whosonfirst-data/master/data/';
-    this.getWOFURL = opts.wof.getWOFURL || this.getDefaultWOFURLFunction();
     this.opts.wof.fillColor = opts.wof.fillColor || "rgba(200, 40, 32, 0.1)";
     this.opts.wof.fillOutlineColor = opts.wof.fillOutlineColor || "rgba(200, 40, 32, 0.7)";
+    this.getWOFURL = opts.wof.getWOFURL || this.getDefaultWOFURLFunction();
   }
+
   if (opts.params) {
     this.params = '';
-    for (var i in opts.params) {
-      this.params += '&' + i + '=' + opts.params[i];
+    for (var key in opts.params) {
+      if (opts.params.hasOwnProperty(key)) {
+        this.params += '&' + key + '=' + opts.params[key];
+      }
     }
   }
+
   this.markerLayerId = 'pelias-mapbox-gl-js-marker';
   this.polygonLayerId = 'pelias-mapbox-gl-js-polygon';
-
-  this._removeMarkers = this._removeSources.bind(this, this.opts.marker, this.markerLayerId);
   this._removePolygon = this._removeSources.bind(this, this.opts.wof, this.polygonLayerId);
+
+  this._keyCodes = {};
+  this._keyCodes.enter = 13;
+  this._keyCodes.arrowUp = 38;
+  this._keyCodes.arrowDown = 40;
 }
 
-PeliasGeocoder.prototype.onAdd = function(map) {
-  var self = this;
+PeliasGeocoder.prototype.onAdd = function (map) {
   this._map = map;
 
-  var el = this.container = document.createElement('div');
-  el.className = 'pelias-ctrl-geocoder mapboxgl-ctrl';
-
-  var icon = document.createElement('span');
-  icon.className = 'geocoder-icon geocoder-icon-search';
+  this._buildWrapper();
+  this._buildIconSearch();
+  this._buildInputWrapper();
+  this._buildInputActionsWrapper();
+  this._buildInput();
+  this._buildResults();
```

Comment:
Two many function call... All the stuff is hidden and harder to read what is really done. (Especially for the creation of HTML elements)